### PR TITLE
Mixed precision training support

### PR DESCRIPTION
**Pull request by godmoves**
Add mixed precision training support. In this method we do all computation in fp16 except loss calculation and the weight update/storage, this will speed up the training procedure and help us use larger batch size.

### CODE REVIEW

**Comment by godmoves** [comment]
Here is the speed test. We can see there is a relatively large speedup for large networks. For GPU with TensorCore, it will use that for the fp16 computation automatically, so the speed improvement for smaller net is larger than the GPU without that.

1080ti (no TensorCore)

|Architecture|Batch Size|FP32 Speed (pos/s)|FP16 Speed (pos/s)|Speed Boost|
|------------|----------|------------------|------------------|-----------|
|10x128      |128       |1300              |1490              |14.6%      |
|            |256       |1195              |1397              |16.9%      |
|            |512       |OOM*              |1358              |           |
|20x256      |128       |205               |337               |64.4%      |
|            |256       |OOM*              |278               |           |

2080ti (TensorCore)

|Architecture|Batch Size|FP32 Speed (pos/s)|FP16 Speed (pos/s)|Speed Boost|
|------------|----------|------------------|------------------|-----------|
|10x128      |128       |1752              |2543              |45.1%      |
|            |256       |1810              |2753              |52.1%      |
|            |512       |OOM*              |2529              |           |
|20x256      |128       |381               |562               |47.5%      |
|            |256       |OOM*              |251               |           |

OOM: Out of memory.

**Comment by gcp** [comment]
I'll try testing it on Windows & with an RTX.

**Comment by roy7** [comment]
Any test results to report @gcp? :)

**Comment by gcp** [comment]
I don't. I no longer manage to get TensorFlow working in Windows at all.

The problem is that I've upgraded to CUDA 10.0 to get Turing features, but the official TensorFlow releases, for god knows what reason, are still on CUDA 9.0 and MSVC2015.

Nightlies do seem to be on CUDA 10.0 but are too broken and crash with obscure error messages.

So I have no good way to test this right now.

**Comment by apetresc** [comment]
> The problem is that I've upgraded to CUDA 10.0 to get Turing features, but the official TensorFlow releases, for god knows what reason, are still on CUDA 9.0 and MSVC2015.

This doesn't appear to be the case anymore since 1.13.1, so maybe it's worth a try now?

**Comment by gcp** [comment]
I am pretty sure it was downloading 1.13.x.

**Comment by godmoves** [comment]
For networks larger than 40b256f, the reg_term of the initial random weight might be too large and get out of the fp16 range. So I cast the l2 loss of each weight back to fp32 before sum them up to prevent the possible overflow.

**Comment by gcp** [comment]
I'll try this again now, the docs seem to really claim CUDA 10 support for latest TensorFlow.

**Comment by vipmath** [comment]
@gcp  https://github.com/fo40225/tensorflow-windows-wheel/tree/master/1.13.1/py37/GPU/cuda101cudnn75sse2

**Comment by gcp** [comment]
I can only confirm what I already said: TensorFlow 1.13.1 does not support CUDA 10.

**Comment by gcp** [comment]
Current TensorFlow Nightlies "work" in the sense that they start up, but they hang after loading the shufflebuffers and nothing seems to happen on the GPU.

TensorFlow 2.0 alpha is not backwards compatible and can't run the code.

**Comment by gcp** [comment]
@vipmath That one won't work (wrong CUDA, wrong Python), but that repo does have a version of 1.12.0 that might be suitable.

**Comment by gcp** [comment]
Works, about 33% speedup on a Turing card.
